### PR TITLE
migrate direct usage of lru to cache2

### PR DESCRIPTION
**Pull request by liusy182**
#### Summary

This PR is a continuation from [14118](https://github.com/mattermost/mattermost-server/pull/14118). It migrates all direct references of the `cache/lru` to the new lru implementation in the `cache2` package.

#### Ticket link

https://mattermost.atlassian.net/browse/MM-25115

### CODE REVIEW

**Comment by agnivade** [comment]
Does this depend on the provider PR going in first ? Or this is indepenent ?

**Comment by liusy182** [comment]
@agnivade this is independent from the provider PR. I will have to take a look of the test failures first before it is ready for review.

**Comment by teresa-novoa** [comment]
@agnivade are you the Dev reviewer for this PR? cc @hanzei Thanks!

**Comment by agnivade** [comment]
It is me and Jesus, yes.

**Comment by agnivade** [comment]
@jespino - Should there be a JIRA for this ? This is a big change and would need a good amount of testing, especially in HA.

**Comment by jespino** [comment]
@agnivade It should, but I think is not still there. We can create one ticket for that, specially for QA to take this changes into consideration.

**Comment by agnivade** [comment]
Done - https://mattermost.atlassian.net/browse/MM-25115

File: api4/openGraph.go [hunk 1]
```diff
@@ -43,15 +46,16 @@ func getOpenGraphMetadata(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	err := openGraphDataCache.Get(url, &ogJSONGeneric)
+	if err == nil {
+		w.Write(ogJSONGeneric)
 		return
 	}
 
 	og := c.App.GetOpenGraphMetadata(url)
 	ogJSON, err := og.ToJSON()
-	openGraphDataCache.AddWithExpiresInSecs(url, ogJSON, 3600) // Cache would expire after 1 hour
+	openGraphDataCache.SetWithExpiry(url, ogJSON, 1*time.Hour) // Cache would expire after 1 hour
```
Comment (agnivade) [inline api4/openGraph.go:58]: Let's remove the comment, since now it's apparent that it's 1 hour.

File: api4/system.go [hunk 2]
```diff
@@ -427,14 +430,14 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
 	res, err := client.Head(url)
 	if err != nil {
 		// Cache failures to prevent retries.
-		redirectLocationDataCache.AddWithExpiresInSecs(url, "", 3600) // Expires after 1 hour
+		redirectLocationDataCache.SetWithExpiry(url, "", 1*time.Hour) // Expires after 1 hour
```
Comment (agnivade) [inline api4/system.go:433]: Same as above. Let's remove the comments from here and everywhere else.

**Comment by agnivade** [comment]
5.24 is cut, so this is ready to be merged. Just waiting on me to run a load test to assess the performance impact. I will get to it soon.

**Comment by agnivade** [comment]
Well, this was certainly interesting. I tested a single mattermost instance with 2k users, and there wasn't any noticeable difference in API response times. If I had to squint very hard, there's probably a 1ms diff in avg time, and probably 2-3ms diff in p90. But that may just be noise.

Here are the graphs

cache
top-cache

total-cache

master

top-master

total-master

Good to go from my side.

**Pull request merged**